diem-move/parallel-executor/src/executor.rs

```diff
@@ -1,235 +1,300 @@
-                                        versioned_data_cache.write(&k, idx, v).is_ok()
-                                    }) {
-                                        scheduler.set_stop_version(idx + 1);
-                                        ExecutionStatus::SkipRest(output)
-                                    } else {
-                                        // Failed to write to the versioned data cache as
-                                        // transaction write to a key that wasn't estimated by the
-                                        // inferencer, aborting the entire execution.
-                                        ExecutionStatus::Abort(Error::UnestimatedWrite)
-                                    }
-                                }
-                                ExecutionStatus::Abort(err) => {
-                                    // Abort the execution with user defined error.
-                                    scheduler.set_stop_version(idx + 1);
-                                    ExecutionStatus::Abort(Error::UserError(err.clone()))
-                                }
-                            };
-
-                        for write in txn_accesses.keys_written.iter() {
-                            // Unwrap here is fine because all writes here should be in the mvhashmap.
-                            assert!(versioned_data_cache.skip_if_not_set(write, idx).is_ok());
+                    let mut scheduler_task = SchedulerTask::NoTask;
+                    loop {
+                        scheduler_task = match scheduler_task {
+                            SchedulerTask::ValidationTask(version_to_validate) => self.validate(
+                                version_to_validate,
+                                &last_input_output,
+                                &versioned_data_cache,
+                                &scheduler,
+                            ),
+                            SchedulerTask::ExecutionTask(version_to_execute) => self.execute(
+                                version_to_execute,
+                                &signature_verified_block,
+                                &last_input_output,
+                                &versioned_data_cache,
+                                &scheduler,
+                                &executor,
+                            ),
+                            SchedulerTask::NoTask => scheduler.next_task(),
+                            SchedulerTask::Done => break,
                         }
-
-                        scheduler.finish_execution(idx);
-                        outcomes.set_result(idx, commit_result);
                     }
                 });
             }
         });
 
-        // Splits the head of the vec of results that are valid
+        // Extract outputs in parallel
         let valid_results_length = scheduler.num_txn_to_execute();
+        (0..valid_results_length)
+            .into_par_iter()
```

Comment:
into_par_chunk is probably more efficient than par_iter but maybe doesn't matter much